src/add_display_name.rs

```diff
@@ -18,42 +18,72 @@ impl Component {
     }
 }
 
-pub struct AddDisplayNameVisitor;
+impl Component {
+    pub fn create_display_name_stmt(&self) -> ModuleItem {
+        ModuleItem::Stmt(
+            Stmt::Expr(ExprStmt {
+                span: DUMMY_SP,
+                expr: Box::new(Expr::Assign(AssignExpr {
+                    span: DUMMY_SP,
+                    op: AssignOp::Assign,
+                    left: PatOrExpr::Expr(Box::new(Expr::Member(MemberExpr {
+                        span: DUMMY_SP,
+                        obj: Box::new(Expr::Ident(Ident::new(self.name.clone(), Span { ctxt: self.ctx, ..DUMMY_SP }))),
+                        prop: MemberProp::Ident(Ident::new(JsWord::from("displayName").into(), DUMMY_SP))
+                    }))),
+                    right: Box::new(Expr::Lit(Lit::Str(Str::from(self.name.clone()))))
+                }))
+            })
+        )
+    }
+}
+
+pub struct AddDisplayNameVisitor {
+    components: Vec<Component>,
+}
+
+impl Default for AddDisplayNameVisitor {
+    fn default() -> AddDisplayNameVisitor {
+        AddDisplayNameVisitor {
+            components: Vec::new(),
+        }
+    }
+}
 
 impl VisitMut for AddDisplayNameVisitor {
     fn visit_mut_module_items(&mut self, stmts: &mut Vec<ModuleItem>) {
         stmts.visit_mut_children_with(self);
 
-        let mut components: Vec<Component> = Vec::new();
-        stmts.iter_mut().enumerate().for_each(|(i, stmt)| {
-            if let Some(comp) = export_var_decl(stmt) { components.push(comp.with_pos(i)) }
-            if let Some(comp) = var_decl_stmt(stmt) { components.push(comp.with_pos(i)) }
-            if let Some(comp) = default_export_fn_decl(stmt) { components.push(comp.with_pos(i)) }
-            if let Some(comp) = export_fn_decl(stmt) { components.push(comp.with_pos(i)) }
-            if let Some(comp) = bare_fn_decl(stmt) { components.push(comp.with_pos(i)) }
-        });
-
-        components.iter().enumerate().for_each(|(i, comp)| {
+        self.components.iter().enumerate().for_each(|(i, comp)| {
             let index = i + comp.pos + 1;
-            stmts.insert(index, ModuleItem::Stmt(set_display_name_stmt(comp)));
+            if index < stmts.len() {
+                stmts.insert(index, comp.create_display_name_stmt());
+            } else {
+                stmts.push(comp.create_display_name_stmt());
+            }
         })
     }
-}
 
-fn export_var_decl(stmt: &mut ModuleItem) -> Option<Component> {
-    let var_decls = stmt.as_mut_module_decl()?.as_mut_export_decl()?.decl.as_mut_var()?.as_mut();
-    process_var_decls(var_decls)
-}
+    fn visit_mut_var_declarator(&mut self, n: &mut VarDeclarator) {
+        if let Some(comp) = process_var_declarator(n) {
```

Comment:
Is `n.visit_mut_children_with(self);` intentionally omitted? Thus only top-level statements will be visited.

Reply:
@hjkcai Not intentionally. Should I add it back?

Reply:
@hjkcai I don't think we should add it. Here's what I've tried:

1. Added a failing test case that supports nested var:

```rs
    test!(SYNTAX, runner,
        /* Name */ nested_var,
        /* Input */ r#"
            const Foo = () => <div />;
            export const Bar = (function iife() {
                var Baz = () => <article />;
                return Baz;
            })();
        "#,
        /* Output */ r#"
            const Foo = () => <div />;
            Foo.displayName = "Foo";
            export const Bar = function iife() {
                var Baz = ()=><article/>;
                Baz.displayName = "Baz";
                return Baz;
            }();
            Bar.displayName = "Bar";
        "#
    );
```

2. Added `n.visit_mut_children_with(self);` to three visitor functions
3. Then I get unexpected results

![JPEG-1](https://github.com/hjkcai/swc-plugin-add-display-name/assets/922234/4fdedd22-45dc-4280-b5fc-6d3f8bf829be)
![JPEG-2](https://github.com/hjkcai/swc-plugin-add-display-name/assets/922234/8a7cbbae-cf4d-49f6-9980-f466a7234894)

From what I've read of your original code, I don't think we need to support **non-top-level** yet. I suggest we keep this as is.